dask/array/core.py

```diff
@@ -889,65 +889,61 @@ def store(sources, targets, lock=True, regions=None, compute=True,
     sources_dsk = sharedict.merge(*[e.__dask_graph__() for e in sources])
     sources_dsk = Array.__dask_optimize__(
         sources_dsk,
-        [e.__dask_keys__() for e in sources]
+        list(core.flatten([e.__dask_keys__() for e in sources]))
     )
 
-    tgt_dsks = []
-    store_keys = []
-    store_dsks = []
-    if return_stored:
-        load_names = []
-        load_dsks = []
-    for tgt, src, reg in zip(targets, sources, regions):
-        # if out is a delayed object update dictionary accordingly
+    # Optimize all targets together
+    targets_keys = []
+    targets_dsk = []
+    for e in targets:
         try:
-            each_tgt_dsk = {}
-            each_tgt_dsk.update(tgt.dask)
-            tgt = tgt.key
+            targets_keys.extend(e.__dask_keys__())
+            targets_dsk.append(e.__dask_graph__())
         except AttributeError:
-            each_tgt_dsk = {}
+            pass
+
+    targets_dsk = sharedict.merge(*targets_dsk)
+    targets_dsk = Delayed.__dask_optimize__(targets_dsk, targets_keys)
```

Comment:
Can I ask you to provide a small example where this has a significant positive impact?  In general going between sharedicts and dicts or calling optimizations like cull are not free.

Reply:
Ah, I'm sorry, this is for the targets.  Should these be co-optimized with the sources?  Otherwise if they share common histories then that common history will be lost.

Reply:
by lost I mean redundantly computed.

Reply:
Not sure I follow. The `sources` are Dask Arrays and the `targets` are Dask Delayed objects. How would one optimize them together? Presumably through some function instead of the `__dask_optimize__` method. Would that other function still optimize the Dask Arrays correctly?

As a side note, IIUC `Delayed.__dask_optimize__` is a no-op currently. So I don't think this does anything beyond accumulate all the target Dasks into one `SharedDict` for later. That said, thought we should try to do the right thing here should `Delayed.__dask_optimize__`'s behavior ever change in the future. Would add that if a change to `Delayed.__dask_optimize__` did occur, it would be fair to evaluate its effects here and anywhere it might have impact.

Reply:
Probably should also add that the main use case for `Delayed` targets was to allow one to create a storage location only once it was needed. So expect in practice these are quite shallow and don't share anything with the Dask Arrays sources. IOW probably not worth worrying about unless someone brings up a use case.